uefi/src/proto/network/snp.rs

```diff
@@ -0,0 +1,587 @@
+        (self.stop)(self).into()
+    }
+
+    /// Resets a network adapter and allocates the transmit and receive buffers
+    /// required by the network interface; optionally, also requests allocation of
+    /// additional transmit and receive buffers
+    pub fn initialize(&self, extra_rx_buffer_size: usize, extra_tx_buffer_size: usize) -> Result {
+        (self.initialize)(self, extra_rx_buffer_size, extra_tx_buffer_size).into()
+    }
+
+    /// Resets a network adapter and reinitializes it with the parameters that were
+    /// provided in the previous call to `initialize`
+    pub fn reset(&self, extended_verification: bool) -> Result {
+        (self.reset)(self, extended_verification).into()
+    }
+
+    /// Resets a network adapter and leaves it in a state that is safe
+    /// for another driver to initialize
+    pub fn shutdown(&self) -> Result {
+        (self.shutdown)(self).into()
+    }
+
+    /// Manages the multicast receive filters of a network
+    pub fn receive_filters(
+        &self,
+        enable: u32,
+        disable: u32,
+        reset_mcast_filter: bool,
+        mcast_filter: Option<&[MacAddress]>,
+    ) -> Result {
+        if let Some(mcast_filter) = mcast_filter {
+            (self.receive_filters)(
+                self,
+                enable,
+                disable,
+                reset_mcast_filter,
+                mcast_filter.len(),
+                NonNull::new(mcast_filter.as_ptr() as *mut _),
+            )
+            .into()
+        } else {
+            (self.receive_filters)(self, enable, disable, reset_mcast_filter, 0, None).into()
+        }
+    }
+
+    /// Modifies or resets the current station address, if supported
+    pub fn station_address(&self, reset: bool, new: Option<&MacAddress>) -> Result {
+        (self.station_address)(self, reset, new).into()
+    }
+
+    /// Resets statistics on a network interface
+    pub fn reset_statistics(&self) -> Result {
+        (self.statistics)(self, true, None, None).into()
+    }
+
+    /// Collects statistics on a network interface
+    pub fn collect_statistics(&self) -> Result<NetworkStats> {
+        let mut stats_table: NetworkStats = Default::default();
+        let mut stats_size = core::mem::size_of::<NetworkStats>();
+        let status = (self.statistics)(self, false, Some(&mut stats_size), Some(&mut stats_table));
+        Result::from(status)?;
+        Ok(stats_table)
+    }
+
+    /// Converts a multicast IP address to a multicast HW MAC Address
+    pub fn mcast_ip_to_mac(&self, ipv6: bool, ip: IpAddress) -> Result<MacAddress> {
+        let mut mac_address = MacAddress([0; 32]);
+        let status = (self.mcast_ip_to_mac)(self, ipv6, &ip, &mut mac_address);
+        Result::from(status)?;
+        Ok(mac_address)
+    }
+
+    /// Performs read operations on the NVRAM device attached to
+    /// a network interface
+    pub fn read_nv_data(&self, offset: usize, buffer_size: usize, buffer: *mut c_void) -> Result {
```

Comment:
For the public interface I think we want a slice for the buffer; `&mut [u8]` for `read_nv_data` and `&[u8]` for `write_nv_data`.